### PR TITLE
Fixes for gpt-oss

### PR DESCRIPTION
**Pull request by cuichenx**
# What does this PR do ?
Various minor fixes for gpt-oss model
1. Fix yarn mscale to match with [HF implementation](https://github.com/huggingface/transformers/blob/v4.57-release/src/transformers/modeling_rope_utils.py#L324) and disable rope fusion when mscale is used.
2. Initialize softmax offset on GPU
3. Remove lru cache for attention mask to prevent OOM with long sequence lengths
Works with https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1155

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
4. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by cuichenx** [comment]
/ok to test [c5b1e1e](https://github.com/NVIDIA/Megatron-LM/pull/2038/commits/c5b1e1e8542272f5001480530c4db5bbb2b11a27)

File: megatron/core/models/common/embeddings/yarn_rotary_pos_embedding.py
Comment (ashors1) [inline megatron/core/models/common/embeddings/yarn_rotary_pos_embedding.py]: We only end up here if "yarn_mscale", and"yarn_mscale_all_dim" are defined. Don't we want to call `_yarn_get_concentration_factor` even if `yarn_rotary_scaling_factor` is set but `config.yarn_mscale` and `config.yarn_mscale_all_dim` are not?

Comment (cuichenx) [reply megatron/core/models/common/embeddings/yarn_rotary_pos_embedding.py]: True, that's probably the best.

File: megatron/core/transformer/utils.py [hunk 1]
```diff
@@ -29,13 +29,11 @@ def get_linear_layer(rows, columns, init_method, perform_initialization=True):
     return layer
 
 
-@lru_cache(maxsize=32)
 def get_default_causal_mask(sq: int) -> torch.Tensor:
     """Return the causal upper triangular mask for softmax input."""
     return torch.triu(torch.ones(sq, sq, device="cuda"), diagonal=1).bool()
 
 
-@lru_cache(maxsize=32)
```
Comment (cuichenx) [inline megatron/core/transformer/utils.py:38]: leads to OOM when seqlen is large.

File: megatron/core/models/common/embeddings/yarn_rotary_pos_embedding.py [hunk 2]
```diff
@@ -228,22 +228,25 @@ def _yarn_get_mscale(scale: float = 1, mscale: float = 1) -> float:
...
 
 def _yarn_get_concentration_factor_from_config(config: TransformerConfig) -> float:
-    fields = ["yarn_rotary_scaling_factor", "yarn_mscale", "yarn_mscale_all_dim"]
-    if all(hasattr(config, f) for f in fields):
+    if hasattr(config, "yarn_rotary_scaling_factor"):
         return _yarn_get_concentration_factor(
-            config.yarn_rotary_scaling_factor, config.yarn_mscale, config.yarn_mscale_all_dim
+            config.yarn_rotary_scaling_factor,
+            getattr(config, "yarn_mscale", None),
+            getattr(config, "yarn_mscale_all_dim", None),
```
Comment (cuichenx) [inline megatron/core/models/common/embeddings/yarn_rotary_pos_embedding.py:250]: matches HF implementation https://github.com/huggingface/transformers/blob/v4.57-release/src/transformers/modeling_rope_utils.py#L324

**Comment by cuichenx** [comment]
/ok to test [eedb6bb](https://github.com/NVIDIA/Megatron-LM/pull/2038/commits/eedb6bbba7e85910653ba7a6c3f7ca7262af5de7)

**Comment by cuichenx** [comment]
/ok to test [de2ec7c](https://github.com/NVIDIA/Megatron-LM/pull/2038/commits/de2ec7cb338d9fe815da0fd2efe8e203c287d399)

**Comment by cuichenx** [comment]
/ok to test [f763cd3](https://github.com/NVIDIA/Megatron-LM/pull/2038/commits/f763cd33430ce625eceb7f2c17b1676409ec441c)

**Comment by kvareddy** [comment]
@santhnm2, can you please take a look at this MR?

**Review by santhnm2 — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]
